### PR TITLE
[feat] Implement task management system functionality

### LINKED ISSUE #95: Implement task management system functionality
https://linear.app/trap-garden/issue/TG-66/implement-task-management-system-functionality

### PR DESCRIPTION
**Pull request by SashankBalusu**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
The general premise of this sprint was basically to allow the users to properly track harvesting, weeding, and watering tasks. 
Without going into the details here are the general implementations: 
1. The user will get weekly reminders on when to water. We redisplay the task when there are three days left until one full week has elapsed since the last watering. 
2. For weeding, the user either gets biweekly or weekly tasks. Weekly tasks reappear similar to watering tasks, biweekly tasks reappear after 10 days
3. For harvesting, the task appears by season (there's special logic for the winter season that occurs across years). If the task is marked completed, the plant also gets harvested and the associated logic plays out (as long as the user confirms w the confirmation modal)
4. A confirmation modal appears whenever a user unchecks a task. 
5. Note: the confirmation modal's buttons are weirdly worded rn, i will fix this by passing in button text as props if this isnt already what Torin's sprint is doing

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! Use the provided image template. Drag the desired image into the PR, then copy the link into the placeholder."
image
image
image
image

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
1. There are a ton of Supabase columns added with this sprint. IIRC they are plant_name, harvest_season, last_watered, last_weeded, due_date, date_added_to_db, water_frequency, and weeding_frequency, previous_last_watered, and previous_last_weeded
2. Look through the back end functions in dashboard and userPlants
3. Basically all of the logic changes take place in dashboard. There are some comments but honestly the best way to review is probably just to mess around with the code and then just ask me to walk u thru the file bc its a bit of a mess and some of the logic is weird at times!
4. the add details updates are purely to ensure that the database gets the correct information into userplants for the new rows added

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
As Kyle mentioned, it might be better to have a separate tasks table instead of all the additional columns on user plants

### Online sources
[//]: # 'Copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs
[//]: # "Add related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @kylezryr

### CODE REVIEW

File: api/supabase/queries/dashboard.ts [hunk 1]
```diff
@@ -42,3 +42,64 @@ export const getDailyPlantTip = async (): Promise<PlantTip | null> => {
...
+  return data ?? [];
+}
+
+export async function updateDate(
+  id: string,
+  newDate: Date,
+  taskType: string,
+  isPrevious: boolean = false,
+) {
+  console.log(id, newDate, taskType, isPrevious);
```
Comment (kylezryr) [inline api/supabase/queries/dashboard.ts]: remove console log

File: api/supabase/queries/dashboard.ts [hunk 2]
```diff
@@ -42,3 +42,64 @@ export const getDailyPlantTip = async (): Promise<PlantTip | null> => {
...
+    .from('user_plants')
+    .update({ due_date: new_date })
+    .eq('id', id);
+
+  if (error) {
+    console.error(`Error updating due date:`, error);
+  }
+
+  return data;
+}
```
Comment (kylezryr) [inline api/supabase/queries/dashboard.ts]: functions look good!

File: app/dashboard/styles.ts
Comment (kylezryr) [inline app/dashboard/styles.ts:1]: image
got this error when first adding plants to my garden then loading the dashboard. try to replicate by having no plants in your garden, adding a few (i added carrots (summer), cabbage, and lettuce (summer)). i think this is generated by the styled components `active` prop in dashboard, i got a similar error for `completed` too

File: app/dashboard/page.tsx [hunk 3]
```diff
@@ -1,44 +1,386 @@
...
+    }>
+  >([]);
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  // Use one unified modal action state
+  const [modalAction, setModalAction] = useState<ModalAction | null>(null);
+  const [selectedTab, setSelectedTab] = useState<'current' | 'completed'>(
+    'current',
+  );
+
+  // Helper functions for comparing dates
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: can we move all of these helper functions to a new `taskHelpers.ts` file in `utils`, i don't wanna overcrowd the dashboard page since these functions aren't returning any React, plus these are a lot of functions that probably aren't necessary to be here

File: app/dashboard/page.tsx [hunk 4]
```diff
@@ -1,44 +1,386 @@
...
+
+  const isOlderThanWateringFrequencyOrNull = (date: string | null): boolean => {
+    if (date == null) return true;
+    const givenDate = new Date(date);
+    const currentDate = new Date();
+    currentDate.setDate(currentDate.getDate() - 3);
+    return givenDate < currentDate;
+  };
+
+  const isOlderThanWeedingFrequencyOrNull = (
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: can we be consistent with either using arrow functions (`const funcName = () => {}`) or `function`? since the rest of the functions in helpers use `function`, lets use that too for these helper functions instead of arrow function definitions

File: app/dashboard/page.tsx [hunk 5]
```diff
@@ -1,44 +1,386 @@
...
+    'current',
+  );
+
+  // Helper functions for comparing dates
+
+  const isOlderThanWateringFrequencyOrNull = (date: string | null): boolean => {
+    if (date == null) return true;
+    const givenDate = new Date(date);
+    const currentDate = new Date();
+    currentDate.setDate(currentDate.getDate() - 3);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: could u add a comment as to why -3? if im correct this is because we want to show the watering task after 3 days?

File: app/dashboard/page.tsx [hunk 6]
```diff
@@ -1,44 +1,386 @@
...
+        );
+      } else if (currentMonth === 11) {
+        // December
+        // Only accept if the harvest was in December of the current year.
+        return (
+          recent.getFullYear() === current.getFullYear() &&
+          recent.getMonth() === 11
+        );
+      }
+      // This fallback shouldn't be reached if all WINTER months are covered
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: god i hope this if statement logic is correct, i cannot be asked to check it manually

File: app/dashboard/page.tsx [hunk 7]
```diff
@@ -1,44 +1,386 @@
...
+    } else {
+      candidateDueDate = new Date();
+      candidateDueDate.setDate(candidateDueDate.getDate() + interval);
+    }
+    setDueDate(candidateDueDate, taskId);
+    return candidateDueDate;
+  };
+
+  // Process tasks from the database (UserPlant[]) and add water, weed, and harvest tasks.
+  function getValidTasks(tasks: UserPlant[]) {
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: leave this function in dashboard, no need to move to helpers

File: app/dashboard/page.tsx [hunk 8]
```diff
@@ -1,44 +1,386 @@
...
+      type: 'water' | 'weed' | 'harvest';
+      plant_name: string;
+      completed: boolean;
+      due_date: Date;
+      id: string;
+      // For water/weed tasks we keep a previousDate
+      previousDate?: Date;
+      // For harvest tasks, store the season and an optional due message.
+      harvestSeason?: string;
+      dueMessage?: string;
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: should we define a new type for validTasks? might be better to enforce typechecking later on

File: app/dashboard/page.tsx [hunk 9]
```diff
@@ -1,44 +1,386 @@
...
+        validTasks.push({
+          type: 'water',
+          plant_name: task.plant_name,
+          completed: false,
+          due_date: due,
+          id: task.id,
+          previousDate: new Date(task.last_watered),
+        });
+        if (task.previous_last_watered !== task.last_watered) {
+          updateDate(task.id, new Date(task.last_watered), 'water', true);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: this is a supabase call right? if so we should `await` and make `getValidTasks` an async function

File: app/dashboard/page.tsx [hunk 10]
```diff
@@ -1,44 +1,386 @@
...
 
+  useEffect(() => {
+    if (!authLoading && userId) {
+      const fetchTasks = async () => {
+        const tasks = await getPendingTasks(userId);
+        setPendingTasks(getValidTasks(tasks));
+      };
+      fetchTasks();
+    }
+  }, [authLoading, userId]);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: add getValidTasks as a dependency (lint warning)

File: app/dashboard/page.tsx [hunk 11]
```diff
@@ -1,44 +1,386 @@
...
+
+  // Filter tasks based on the selected tab.
+  const filteredTasks = pendingTasks.filter(task =>
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: we should put this in a useMemo with dependency of selectedTab to avoid recomputation which can be slow, see planting timeline page for an example of useMemo

File: app/dashboard/page.tsx [hunk 12]
```diff
@@ -47,32 +389,88 @@ export default function Page() {
...
+                      if (task.completed) {
+                        handleHarvestClear(task.id);
+                      } else {
+                        handleHarvestSet(task.id);
+                      }
+                    } else {
+                      if (task.completed) {
+                        handleUncheck(task.id, task.type);
+                      } else {
+                        handleCheck(task.id, task.type);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: nit: can we move these if statements into a `handleTaskToggle` function?

**Review by kylezryr — CHANGES_REQUESTED** [review]
holy shit massive PR but everything seems to work apart from the bugs i mentioned. you're gonna have to walk me thru the changes too. no styling comments this time since its the next PR but really good job man this works really well good shit

**Comment by kylezryr** [comment]
another bug: when adding plants and setting date planted to a day in the future (e.g. if today is 17 Apr and I set date platned to be Apr 30), i still see the tasks for that plants as tasks for this week, and the due date is set weird (look at the most recently added plants in this [filtered table](https://supabase.com/dashboard/project/iqcjborfwjjdysukvgel/editor/29743?sort=date_added:desc&filter=user_id:eq:95d28b29-ee37-436a-9193-4fc26266ab2d) ). i think in the future we'll just prevent users from setting date planted to be a date in the future (since this makes no sense anyways) so the bug should be avoided but just to lyk

image

File: app/dashboard/styles.ts
Comment (SashankBalusu) [reply app/dashboard/styles.ts:1]: yea i keep getting this too i lwk dont know how to fix it tho cuz the attribute is literally a boolean in the styled component
lemme look into it

File: app/dashboard/page.tsx (hunk 6 above)
Comment (SashankBalusu) [reply app/dashboard/page.tsx]: lmaoo i tried to hand trace it so ideally its correct but theres a good chance i messed up icl

File: app/dashboard/page.tsx (hunk 8 above)
Comment (SashankBalusu) [reply app/dashboard/page.tsx]: facts

File: types/schema.d.ts [hunk 13]
```diff
@@ -83,3 +93,32 @@ export type TipCategory =
...
+  user_id: UUID;
+  plant_id: UUID;
+  date_added_to_db: string;
+  date_removed: string;
+  plant_name: string;
+  previous_completed_date: string;
+  completed_date: string;
+  frequency: string;
+  isCompleted: boolean;
+  type: string;
```
Comment (kylezryr) [inline types/schema.d.ts]: we should enforce this to be 'water' 'weed' or 'harvest' like for ValidTask

File: api/supabase/queries/tasks.ts [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+  taskId: string,
+  completed: boolean,
+  date: Date | null,
+) {
+  const { error } = await supabase
+    .from('tasks')
+    .update({
+      isCompleted: completed,
+      previous_completed_date: completed ? null : (date?.toISOString() ?? null),
+      completed_date: completed ? date?.toISOString() : null,
```
Comment (kylezryr) [inline api/supabase/queries/tasks.ts]: `toISOString()` causes there to be issues with the timezone, so we should use dayjs instead which should be imported after I merge the bug bash fixes PR, code is shown below

```
 //correctly gets current date without being affected by timezones, in YYYY-MM-DD format
 //if we want the timestamp included just remove the .format
  const getDefaultDate = () => dayjs().format('YYYY-MM-DD');
```

File: api/supabase/queries/tasks.ts [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+  if (error) throw error;
+}
+
+export async function toggleTaskCompleted(
+  taskId: string,
+  makeCompleted: boolean,
+  date: Date | null,
+) {
+  const { error } = await supabase
+    .from('TaskTable')
```
Comment (kylezryr) [inline api/supabase/queries/tasks.ts]: shouldn't this be 'tasks'

File: api/supabase/queries/tasks.ts [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+  if (error) {
+    console.error('Error updating task date:', error);
+    return null;
+  }
+
+  // Supabase returns an array; most callers expect a single row
+  return data?.[0] ?? null;
+}
+
+export async function setDueDate(new_date: Date, id: string) {
```
Comment (kylezryr) [inline api/supabase/queries/tasks.ts]: think this is no longer needed with the refactor right?

File: app/dashboard/page.tsx [hunk 17]
```diff
@@ -68,122 +60,139 @@ export default function Page() {
   );
 
   // Process tasks from the database (UserPlant[]) and add water, weed, and harvest tasks.
-  const getValidTasks = useCallback(async (tasks: UserPlant[]) => {
+  const getValidTasks = useCallback(async (tasks: SingleTask[]) => {
     const validTasks: ValidTask[] = [];
     const currentSeason = getCurrentSeason();
-
+    console.log(tasks);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: remove console log

File: app/dashboard/page.tsx [hunk 18]
```diff
@@ -68,122 +60,139 @@ export default function Page() {
...
+            });
+          } else {
+            if (task.frequency == currentSeason) {
+              // this task is not completed set it back to not completed
+              validTasks.push({
+                type: task.type,
+                plant_name: task.plant_name,
+                completed: false,
+                id: task.id,
+                due_date: new Date(),
```
Comment (kylezryr) [inline app/dashboard/page.tsx:93]: am i right in saying that this `new Date()` is just because tasks needs a date object for due_date, but harvest tasks only show the due message so this isn't needed?

File: app/dashboard/page.tsx [hunk 19]
```diff
@@ -68,122 +60,139 @@ export default function Page() {
...
+              updateCompleted(task.id, false);
+            } else {
+              continue;
+            }
+          }
+        } else {
+          const interval =
+            task.type == 'water' ? 7 : task.frequency == 'weekly' ? 7 : 14;
+          if (isOlderThanFreqeuncyOrNull(task.completed_date, interval)) {
+            console.log('got here');
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: remove

File: app/dashboard/page.tsx [hunk 20]
```diff
@@ -68,122 +60,139 @@ export default function Page() {
...
-          completed: false,
-          due_date: due,
-          id: task.id,
-          previousDate: new Date(task.last_weeded),
-        });
-        if (task.previous_last_weeded !== task.last_weeded) {
-          await updateDate(task.id, new Date(task.last_weeded), 'weed', true);
+            //function to modify completed to false in db
+          } else {
+            console.log('got here');
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: remove

File: app/dashboard/page.tsx [hunk 21]
```diff
@@ -68,122 +60,139 @@ export default function Page() {
...
-          completed,
-          due_date: new Date(),
-          dueMessage: `Due end of ${task.harvest_season} season`,
-          id: task.id,
-          harvestSeason: task.harvest_season,
-        });
       }
     }
 
+    console.log(validTasks);
```
Comment (kylezryr) [inline app/dashboard/page.tsx]: remove

File: components/YourPlantDetails/index.tsx [hunk 22]
```diff
@@ -55,10 +60,17 @@ export default function YourPlantDetails({
 
   async function harvestPlant() {
     await increaseHarvestedByOne(id);
-    const currentDate = new Date().toISOString();
-    await setRecentHarvestDate(currentDate, id);
-
-    setLocalRecentHarvestDate(currentDate);
+    const currentDate = new Date();
+    await setRecentHarvestDateThroughId(currentDate.toISOString(), id);
```
Comment (kylezryr) [inline components/YourPlantDetails/index.tsx]: again use dayjs here instead of toISOString to avoid timezone issues

File: utils/taskHelpers.ts [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        recent.getFullYear() === current.getFullYear() &&
+        recent.getMonth() === 11
+      );
+    }
+    // This fallback shouldn't be reached if all WINTER months are covered
+    return recent.getFullYear() === current.getFullYear();
+  }
+}
+
+export function computeDueDate(
```
Comment (kylezryr) [inline utils/taskHelpers.ts:93]: i think there might be an issue with this function. i planted Cilantro on Apr 24 (today), but the due date for watering is 6 days from now instead of 7, and weeding is 13 days instead of 14. I think it may not be adding the interval right? or maybe smth with timezones

image

**Review by kylezryr — CHANGES_REQUESTED** [review]
works pretty well, haven't run into bugs so far. i think the only issue is with how we are currently setting dates using either new Date() or toISOString. we should try to avoid toISOString wherever possible because it converts the datestring into UTC 0 time, affecting the actual date itself. try using dayjs to avoid this. other than that works really well, good job!

**Comment by kylezryr** [comment]
notes for the frontend styling PR:

1. tasks div has white background, should match the rest of the page background
image
2. we can get the user's name now using `userName` from useAuth.
3. add the toast once I merge it too!

File: app/dashboard/page.tsx (hunk 18 above)
Comment (SashankBalusu) [reply app/dashboard/page.tsx:93]: yuh

File: utils/taskHelpers.ts (hunk 23 above)
Comment (SashankBalusu) [reply utils/taskHelpers.ts:93]: i think its cuz of the timezones yea : (

Comment (SashankBalusu) [reply utils/taskHelpers.ts:93]: just confirmed it should be good now
